### PR TITLE
Eiffel2.0 events links validation based on the Bordeaux edition of protocol

### PR DESCRIPTION
**Pull request by xumakap**

### CODE REVIEW

File: src/main/java/com/ericsson/eiffel/remrem/semantics/config/LinksConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.ResourceBundle;
+
+/**
+ * This class is used to read required and optional links from linksValidation.properties file
+ * 
+ * @author xumakap
```
Comment (xmairaa) [inline src/main/java/com/ericsson/eiffel/remrem/semantics/config/LinksConfiguration.java]: remove author

File: src/main/java/com/ericsson/eiffel/remrem/semantics/config/LinksConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    private static final String REQUIRED_LINKS = "requiredLinks";
+    private static final String OPTIONAL_LINKS = "optionalLinks";
+    static HashMap<String, List<String>> requiredLinksMap = new HashMap<String, List<String>>();
+    static HashMap<String, List<String>> optionalLinksMap = new HashMap<String, List<String>>();
+
+    public static List<String> getRequiredLinks(String eventType) {
+
+        if (requiredLinksMap.size() == 0 || optionalLinksMap.size() == 0) {
+            while (keys.hasMoreElements()) {
```
Comment (xmairaa) [inline src/main/java/com/ericsson/eiffel/remrem/semantics/config/LinksConfiguration.java]: Why iterating map, Use directly get from Map

File: src/main/java/com/ericsson/eiffel/remrem/semantics/validator/EiffelValidator.java [hunk 3]
```diff
@@ -67,4 +85,62 @@ public void validate(JsonObject jsonObjectInput) throws EiffelValidationExceptio
...
 
+    public void linksValidation(EiffelEventType eiffelType, JsonObject jsonObject) throws EiffelValidationException {
+        Map<String, Integer> map = new HashMap<String, Integer>();
+        try {
+            List<String> requiredLinks = LinksConfiguration.getRequiredLinks(eiffelType.getEventName());
+            List<String> optionalLinks = LinksConfiguration.getOptionalLinks(eiffelType.getEventName());
+
+            if(requiredLinks == null && optionalLinks == null) {
+                throw new EiffelValidationException(eiffelType.getEventName()+"event links not configured in property file "+CONFIG_FILE);
+            }
```
Comment (xmairaa) [inline src/main/java/com/ericsson/eiffel/remrem/semantics/validator/EiffelValidator.java]: if may be some of event does not have optional parameters then?

File: src/main/java/com/ericsson/eiffel/remrem/semantics/validator/EiffelValidator.java [hunk 4]
```diff
@@ -67,4 +85,62 @@ public void validate(JsonObject jsonObjectInput) throws EiffelValidationExceptio
         }
     }
```
Comment (xmairaa) [inline src/main/java/com/ericsson/eiffel/remrem/semantics/validator/EiffelValidator.java]: Write method doc comments 
http://www.oracle.com/technetwork/articles/java/index-137868.html

File: src/main/java/com/ericsson/eiffel/remrem/semantics/validator/EiffelValidator.java [hunk 5]
```diff
@@ -67,4 +85,62 @@ public void validate(JsonObject jsonObjectInput) throws EiffelValidationExceptio
...
 
+    public void linksValidation(EiffelEventType eiffelType, JsonObject jsonObject) throws EiffelValidationException {
```
Comment (xmairaa) [inline src/main/java/com/ericsson/eiffel/remrem/semantics/validator/EiffelValidator.java]: We are validating only Links Object, So do not pass entire event object.

File: src/main/java/com/ericsson/eiffel/remrem/semantics/validator/OutputValidator.java [hunk 6]
```diff
@@ -14,11 +14,13 @@
...
+import java.util.List;
+
 /**
  * Created by ehenkan on 5/27/16.
  */
 public class OutputValidator extends EiffelValidator {
-    public OutputValidator(String schemaResourceName) {
-        super(schemaResourceName);
+    public OutputValidator(String schemaResourceName,String eventType, List<String> requiredLinks, List<String> optionalLinks, List<String> linkTypes) {
+        super(schemaResourceName, eventType, requiredLinks, optionalLinks, linkTypes);
```
Comment (xvinosi-github) [inline src/main/java/com/ericsson/eiffel/remrem/semantics/validator/OutputValidator.java]: allLinkTypes

File: src/main/java/com/ericsson/eiffel/remrem/semantics/validator/EiffelValidator.java [hunk 7]
```diff
@@ -58,13 +77,64 @@ public void validate(JsonObject jsonObjectInput) throws EiffelValidationExceptio
...
+            }
+            count++;
+            linksCountMapForEvent.put(linkType, count);
+        }
+        Set<String> linksSet = new HashSet<>(linksCountMapForEvent.keySet());
+
+        for (String requiredLink : requiredLinks) {
+            Integer count = linksCountMapForEvent.get(requiredLink);
+            if (count == null) {
+                throw new EiffelValidationException("Mandatory links missing");
```
Comment (xmairaa) [inline src/main/java/com/ericsson/eiffel/remrem/semantics/validator/EiffelValidator.java]: Add Link type to EiffelValidationException

File: src/main/java/com/ericsson/eiffel/remrem/semantics/config/LinksConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        List<String> linkTypeList = new ArrayList<String>();
+        try {
+            if (!links.getString(key).isEmpty()) {
+                linkTypeList = Arrays.asList(links.getString(key).split(","));
+            }
+            return linkTypeList;
+        } catch (MissingResourceException e) {
+            return linkTypeList;
+        }
+    }*/
```
Comment (xvinosi-github) [inline src/main/java/com/ericsson/eiffel/remrem/semantics/config/LinksConfiguration.java]: remove not needed code

File: src/main/java/com/ericsson/eiffel/remrem/semantics/config/LinksConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+     * @param eventType of an Eiffel event
+     *          For Eg: eiffelartifactpublished, eiffelactivityfinished
+     * @return list of optional link types for an event else return an empty list
+     */
+    public List<String> getOptionalLinkTypes(String eventType) {
+        String key = eventType + DOT + OPTIONAL_LINKS;
+        return getLinkTypesFromConfiguration(key);
+    }
+
+    /*private List<String> getLinkTypesFromConfiguration(String key) {
```
Comment (xvinosi-github) [inline src/main/java/com/ericsson/eiffel/remrem/semantics/config/LinksConfiguration.java]: remove commented code

File: src/main/java/com/ericsson/eiffel/remrem/semantics/config/LinksConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+     * @param eventType of an Eiffel event
+     *          For Eg: eiffelartifactpublished, eiffelactivityfinished
+     * @return list of link types required for an event else return an empty list
+     */
+    public List<String> getRequiredLinkTypes(String eventType) {
+        String key = eventType + DOT + REQUIRED_LINKS;
+        return getLinkTypesFromConfiguration(key);
+    }
+
+    private List<String> getLinkTypesFromConfiguration(String key) {
```
Comment (xvinosi-github) [inline src/main/java/com/ericsson/eiffel/remrem/semantics/config/LinksConfiguration.java]: add java doc

**Comment by xvinosi-github** [comment]
+1

File: src/main/java/com/ericsson/eiffel/remrem/semantics/validator/EiffelValidator.java [hunk 11]
```diff
@@ -67,4 +84,62 @@ public void validate(JsonObject jsonObjectInput) throws EiffelValidationExceptio
...
+                Integer count = linksCountMapForEvent.get(linkType);
+                if (count == null) {
+                    count = 0;
+                }
+                count++;
+                linksCountMapForEvent.put(linkType, count);
+            }
+            Set<String> linksSet = new HashSet<>(linksCountMapForEvent.keySet());
+
+            for (String requiredLinkType : requiredLinkTypes) {
```
Comment (xmairaa) [inline src/main/java/com/ericsson/eiffel/remrem/semantics/validator/EiffelValidator.java]: use java8 Lambda Expression

File: src/main/java/com/ericsson/eiffel/remrem/semantics/validator/EiffelValidator.java [hunk 12]
```diff
@@ -67,4 +84,62 @@ public void validate(JsonObject jsonObjectInput) throws EiffelValidationExceptio
...
+     * @throws EiffelValidationException if any validation fails
+     */
+    public void customValidation(JsonObject jsonObjectInput) throws EiffelValidationException {
+        // Links validation
+        Map<String, Integer> linksCountMapForEvent = new HashMap<String, Integer>();
+        String CAUSE_LINK = "CAUSE";
+        String CONTEXT_LINK = "CONTEXT";
+        try {
+            JsonArray links = jsonObjectInput.get("links").getAsJsonArray();
+            for (JsonElement link : links) {
```
Comment (xmairaa) [inline src/main/java/com/ericsson/eiffel/remrem/semantics/validator/EiffelValidator.java]: please use java 8 Lambda Expression, we can reduce this logic.

File: src/main/java/com/ericsson/eiffel/remrem/semantics/validator/EiffelValidator.java [hunk 13]
```diff
@@ -67,4 +84,62 @@ public void validate(JsonObject jsonObjectInput) throws EiffelValidationExceptio
...
+                    throw new EiffelValidationException(
+                            "Multiple trace links are not allowed for link type " + requiredLinkType);
+                }
+                linksSet.remove(requiredLinkType);
+            }
+            if (linksCountMapForEvent.containsKey(CAUSE_LINK) && linksCountMapForEvent.containsKey(CONTEXT_LINK)) {
+                throw new EiffelValidationException(
+                        "Link types " + CAUSE_LINK + " and " + CONTEXT_LINK + " should not be in one event");
+            }
+            for (String optionalLinkType : optionalLinkTypes) {
```
Comment (xmairaa) [inline src/main/java/com/ericsson/eiffel/remrem/semantics/validator/EiffelValidator.java]: use java 8 Lambda Expression

File: src/main/java/com/ericsson/eiffel/remrem/semantics/validator/EiffelValidator.java (hunk 12 above)
Comment (SantoshNC68) [inline src/main/java/com/ericsson/eiffel/remrem/semantics/validator/EiffelValidator.java]: I guess we can implement then in upcoming stories related to this.

**Pull request merged**